src/main/java/com/marklogic/client/ext/document/DocumentWriteOperationBuilder.java

```diff
@@ -15,32 +17,62 @@
 	private String permissions;
 
 	private ContentIdExtractor contentIdExtractor = new DefaultContentIdExtractor();
+	private String idStrategy;
+	private String path;
+	
 
-	public DocumentWriteOperation build(AbstractWriteHandle content, DocumentMetadataHandle metadata ) {
+	public DocumentWriteOperation build(AbstractWriteHandle content, DocumentMetadataHandle metadata, String topic, Integer partition, long offset ) throws IOException {
 		if (content == null) {
 			throw new NullPointerException("'content' must not be null");
 		}
-
 		if (hasText(collections)) {
 			metadata.getCollections().addAll(collections.trim().split(","));
 		}
 		if (hasText(permissions)) {
 			new DefaultDocumentPermissionsParser().parsePermissions(permissions.trim(), metadata.getPermissions());
 		}
-
-		String uri = buildUri(content);
-		return build(operationType, uri, metadata, content);
-	}
-
-	protected String buildUri(AbstractWriteHandle content) {
-		String uri = contentIdExtractor.extractId(content);
+		String uri = "";
+		if ("JSONPATH".equals(idStrategy)) {
```

Comment:
Any time I see a big if/else or switch like this, I think a better approach is to have a separate implementation class for each strategy - https://en.wikipedia.org/wiki/Strategy_pattern - and then have (in this case) a Map<String, ContentIdExtractor> where each key is one of the strings here. 

That forces us to rethink the ContentIdExtractor interface. Instead of adding methods to it, what we want are multiple implementations. 

But, some of the implementations need more information than just the AbstractWriteHandle - they need the topic, partition, and offset. Those are all on the SinkRecord - so add the SinkRecord as an argument to what should be the only method in the ContentIdExtractor interface - e.g. extractId(content, sinkRecord). 

Additionally, the JSONPATH and HASH strategies need one or more paths. Since those don't vary on every call - like the content and sinkRecord do - those should be constructor arguments, not interface method arguments. 

Thus, what DefaultSinkRecordConverter should do is read properties from MarkLogicSinkConfig and use those to build a Map<String, ContentIdExtractor> (which is essentially a lightweight Factory class). That Map is then passed to DocumentWriteOperationBuilder (instead of a single ContentIdExtractor), which will then choose which ContentIdExtractor to use based on the nature of the content... although see below on that, I'm wondering how to avoid those implementation details from leaking out.